lib/executor/constant_arrival_rate.go

```diff
@@ -159,44 +165,49 @@ func (carc ConstantArrivalRateConfig) GetExecutionRequirements(es *lib.Execution
 func (carc ConstantArrivalRateConfig) NewExecutor(
 	es *lib.ExecutionState, logger *logrus.Entry,
 ) (lib.Executor, error) {
-	return ConstantArrivalRate{
+	return &ConstantArrivalRate{
 		BaseExecutor: NewBaseExecutor(carc, es, logger),
 		config:       carc,
 	}, nil
 }
 
 // HasWork reports whether there is any work to be done for the given execution segment.
-func (carc ConstantArrivalRateConfig) HasWork(es *lib.ExecutionSegment) bool {
-	return carc.GetMaxVUs(es) > 0
+func (carc ConstantArrivalRateConfig) HasWork(et *lib.ExecutionTuple) bool {
+	return carc.GetMaxVUs(et) > 0
 }
 
 // ConstantArrivalRate tries to execute a specific number of iterations for a
 // specific period.
 type ConstantArrivalRate struct {
 	*BaseExecutor
 	config ConstantArrivalRateConfig
+	et     *lib.ExecutionTuple
 }
 
 // Make sure we implement the lib.Executor interface.
 var _ lib.Executor = &ConstantArrivalRate{}
 
+// Init values needed for the execution
+func (car *ConstantArrivalRate) Init(ctx context.Context) error {
+	car.et = car.BaseExecutor.executionState.ExecutionTuple.GetNewExecutionTupleBasedOnValue(car.config.MaxVUs.Int64)
+	// TODO mvoe the preallocation of VUs here ?
+	return nil
+}
+
 // Run executes a constant number of iterations per second.
 //
 // TODO: Reuse the variable arrival rate method?
 func (car ConstantArrivalRate) Run(ctx context.Context, out chan<- stats.SampleContainer) (err error) { //nolint:funlen
-	segment := car.executionState.Options.ExecutionSegment
 	gracefulStop := car.config.GetGracefulStop()
 	duration := time.Duration(car.config.Duration.Duration)
-	preAllocatedVUs := car.config.GetPreAllocatedVUs(segment)
-	maxVUs := car.config.GetMaxVUs(segment)
-
-	arrivalRate := getScaledArrivalRate(segment, car.config.Rate.Int64, time.Duration(car.config.TimeUnit.Duration))
+	preAllocatedVUs := car.config.GetPreAllocatedVUs(car.executionState.ExecutionTuple)
+	maxVUs := car.config.GetMaxVUs(car.executionState.ExecutionTuple)
+	arrivalRate := getScaledArrivalRate(car.et.ES, car.config.Rate.Int64, time.Duration(car.config.TimeUnit.Duration))
```

Comment:
For a future PR, but now this `arrivalRate` would be used only for the status text, not for the actual execution, so it's becoming a bit confusing... Please add something like a `//TODO: refactor and simplify` here, since something like this looks more and more appealing for all of the executors: https://github.com/loadimpact/k6/blob/dc46c03ce1a12df30dba8c5b8d103ebc587aaf5e/lib/executor/externally_controlled.go#L366-L369

Reply:
Yeah I also noticed that now there are like 5 variables which are only used for the progressbar ... But definitely another PR at a later stage